### PR TITLE
[TCling] Try more to determine binary path, to locate rdict files (ROOT-10232):

### PR DESCRIPTION
**Pull request by Axel-Naumann**
If the "library" is in fact a binary, it might not have the full path. Instead,
use `/proc/self/exe` or `which` to determine which directory the binary is in.

### CODE REVIEW

File: core/metacling/src/TCling.cxx [hunk 1]
```diff
@@ -1853,7 +1880,9 @@ void TCling::RegisterModule(const char* modulename,
    if (payloadCode)
       code += payloadCode;
 
-   const char *const dyLibName = FindLibraryName(triggerFunc);
+   std::string dyLibNameStr = FindLibraryName(triggerFunc);
+   const char *const dyLibName = dyLibNameStr.empty() ? nullptr : dyLibNameStr.c_str();
```
Comment (vgvassilev) [inline core/metacling/src/TCling.cxx]: Maybe update the uses to `std::string::empty` and drop the char*?

File: core/metacling/src/TCling.cxx [hunk 2]
```diff
@@ -1431,23 +1431,23 @@ void TCling::ShutDown()
 ////////////////////////////////////////////////////////////////////////////////
 /// Wrapper around dladdr (and friends)
 
-static const char *FindLibraryName(void (*func)())
+static std::string FindLibraryName(void (*func)())
 {
 #if defined(__CYGWIN__) && defined(__GNUC__)
    return 0;
```
Comment (amadio) [inline core/metacling/src/TCling.cxx]: This doesn't look like a `std::string` :-)

File: core/metacling/src/TCling.cxx [hunk 3]
```diff
@@ -1456,8 +1456,36 @@ static const char *FindLibraryName(void (*func)())
       // Not in a known share library, let's give up
       return 0;
```
Comment (amadio) [inline core/metacling/src/TCling.cxx]: Idem.

File: core/metacling/src/TCling.cxx [hunk 4]
```diff
@@ -1456,8 +1456,36 @@ static const char *FindLibraryName(void (*func)())
...
    } else {
-      //fprintf(stdout,"Found address in %s\n",info.dli_fname);
+      if (strchr(info.dli_fname, '/'))
+         return info.dli_fname;
+      // Else absolute path. For all we know that's a binary.
+      // Some people have dictionaries in binaries, this is how we find their path:
+      // (see also https://stackoverflow.com/a/1024937/6182509)
+# if defined(R__MACOSX)
+      static char buf[PATH_MAX] = { 0 };
```
Comment (amadio) [inline core/metacling/src/TCling.cxx]: Is this OK not being `thread_local`?

Comment (pcanal) [reply core/metacling/src/TCling.cxx]: In practice it should be okay.  FindLibraryName is static function called only from RegisterModule which should be only called during the initialization part of a library load and this should (ought to?) be protected by the dlopen lock [to be verified though]

File: core/metacling/src/TCling.cxx [hunk 5]
```diff
@@ -1456,8 +1456,35 @@ static const char *FindLibraryName(void (*func)())
...
+      uint32_t bufsize = sizeof(buf);
+      if (_NSGetExecutablePath(buf, &bufsize) >= 0)
+         return buf;
       return info.dli_fname;
+# elif defined(R__LINUX)
+      static char buf[PATH_MAX] = { 0 };
+      // Cross our fingers that /proc/self/exe exists.
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/metacling/src/TCling.cxx]: AFAIK not all distros / unix flavors have `/proc/` so I prefer to keep it the way it is.

File: core/metacling/src/TCling.cxx [hunk 6]
```diff
@@ -1456,8 +1456,35 @@ static const char *FindLibraryName(void (*func)())
...
+      if (!pipe)
+         return info.dli_fname;
+      std::string result;
+      while (fgets(buf, sizeof(buf), pipe)) {
+         result += buffer;
+      }
+      pclose(pipe);
+      return result;
+# else
+#  error "Unsupported platform."
```
[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/metacling/src/TCling.cxx]: Thanks, fixed.

File: core/metacling/src/TCling.cxx (hunk 4 above)
Comment (Axel-Naumann) [reply core/metacling/src/TCling.cxx]: No need for keeping that static anyway. Thanks!

**Comment by Axel-Naumann** [comment]
@phsft-bot build

File: core/metacling/src/TCling.cxx (hunk 5 above)
[the remark this answers is not in the document]
Comment (amadio) [reply core/metacling/src/TCling.cxx]: `R__LINUX` is only defined for Linux, this won't run on other UNIX variants. You need to use `R__UNIX` for that. You are still not testing for the ` (deleted)` case that can happen on Linux.

[the remark this answers is not in the document]
Comment (Axel-Naumann) [reply core/metacling/src/TCling.cxx]: I like that `(deleted)` is added :-) It's certainly not a problem for `TCling`'s usage of this.

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on windows10/default

**Pull request merged**